# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

10 5
15

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
4

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

10 5
15


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

10
5 8 69 2 5 14 63 7 15 2
4347


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [ ]:
import random

for i in range(1,20):
  with open('/content/Test3.txt', 'a') as writefile:
    writefile.write(str(random.randint(2,20))+" ")


In [ ]:
%%time
!./max_pairwise_product_long < Test3.txt

270
CPU times: user 5.89 ms, sys: 4.22 ms, total: 10.1 ms
Wall time: 121 ms


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m8.984s
user	2m8.134s
sys	0m0.031s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product3.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProductFast(const std::vector<int>& numbers) {
    int n = numbers.size();
    int max_index1 = -1;

    for(int i = 0; i < n; ++i){
        if((max_index1 == -1) || (numbers[i]>numbers[max_index1])){
            max_index1 = i;
        }
    }

    int max_index2 = -1;
    for(int j = 0; j<n;++j){
        if((j != max_index1) && ((max_index2 == -1) || (numbers[j]>numbers[max_index2]))){
            max_index2 = j;
        }
    }

    return ((long long) (numbers[max_index1])*numbers[max_index2]);
}

int main() {
    int n;
    //std::cout <<"ingrese n"<< std::endl;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        //std::cout <<"ingrese numbers"<< i << std::endl;
        std::cin >> numbers[i];
    }

    long long result = MaxPairwiseProductFast(numbers);
    std::cout <<result<< "\n";
    return 0;

}


Overwriting max_pairwise_product3.cpp


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product3.cpp -o max_pairwise_product3
!time ./max_pairwise_product3 < input1.txt
!time ./max_pairwise_product3 < input2.txt

90

real	0m0.004s
user	0m0.002s
sys	0m0.001s
40000000000

real	0m0.003s
user	0m0.002s
sys	0m0.001s


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product3 < input3.txt

39999800000

real	0m0.066s
user	0m0.059s
sys	0m0.004s


20 stress tests

In [ ]:
!time ./max_pairwise_product3 < Test3.txt

270

real	0m0.004s
user	0m0.000s
sys	0m0.004s


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product3_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProductFast(const std::vector<int>& numbers, long long &execints) {
    
    long long m = 0LL;
    
    int n = numbers.size();
    execints++;

    int max_index1 = -1;
    execints++;

    execints++;
    for(int i = 0; i < n; ++i){        
        execints += 2;

        execints++;
        if((max_index1 == -1) || (numbers[i]>numbers[max_index1])){
            
            execints++;

            max_index1 = i;
            execints++;
            m++;
        }
        if ( (i <= 10) || (n - i <= 10)) std::cout << m << "\n";
    }

    int max_index2 = -1;

    execints++;
    
    execints++;
    for(int j = 0; j<n;++j){
        execints += 2;
        
        execints++;
        if((j != max_index1) && ((max_index2 == -1) || (numbers[j]>numbers[max_index2]))){
            max_index2 = j;
            execints++;
            m++;
        }
        if ( (j <= 10) || (n - j <= 10)) std::cout << m << "\n";
    }
    return ((long long) (numbers[max_index1])*numbers[max_index2]);
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProductFast(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProductFast(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}

Overwriting max_pairwise_product3_execints.cpp


Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product3_execints.cpp -o max_pairwise_product3_execints
!./max_pairwise_product3_execints < input1.txt
!./max_pairwise_product3_execints < input2.txt
!./max_pairwise_product3_execints < input3.txt

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
19
139
1
2
3
3
3
3
4
5
5
6
7
7
80
1
2
3
4
5
6
7
8
9
10
11
199991
199992
199993
199994
199995
199996
199997
199998
199999
200000
200001
200002
200003
200004
200005
200006
200007
200008
200009
200010
200011
399991
399992
399993
399994
399995
399996
399997
399998
399999
399999
2600009


### Super efficient solution

Without storing the list in memory and without calling a function, add in each line a coment with the number of instructions that are executed in that line 

In [ ]:
%%writefile max_pairwise_product3_super.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    
    int n;                                                                    //1
      
    std::cin >> n;                                                            //1
    std::vector<int> numbers(n);                                              //n
                                                       
    int max_index1 = -1;

    for (int i = 0; i < n; ++i) {                                             //1+2n
        std::cin >> numbers[i];                                               //n
    }

    for(int i = 0; i < n; ++i){                                               //1 + 2n
        if((max_index1 == -1) || (numbers[i]>numbers[max_index1])){           //(n-1)n
            max_index1 = i;                                                       
        }
    }

    int max_index2 = -1;                                                      //1
    for(int j = 0; j < n; ++j){                                               //1+2n
        if((j != max_index1) && ((max_index2 == -1) || (numbers[j]>numbers[max_index2]))){   //(2n-1)n
            max_index2 = j;
        }
    }

    std::cout <<((long long) (numbers[max_index1])*numbers[max_index2])<< "\n"; //3 + 2n + 2[(1+2n)(n-1)n] + (1+2n)
    return 0;

}

Overwriting max_pairwise_product3_super.cpp


Test with inputs 1 and 2


 

In [ ]:
!g++ max_pairwise_product3_super.cpp -o max_pairwise_product3_super
!./max_pairwise_product3_super < input1.txt
!./max_pairwise_product3_super < input2.txt

90
40000000000


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product3_super < input3.txt

39999800000

real	0m0.066s
user	0m0.062s
sys	0m0.001s


20 stress tests 

In [ ]:
!time ./max_pairwise_product3_super < Test3.txt

270

real	0m0.003s
user	0m0.002s
sys	0m0.001s


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product3_super_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    
    long long m = 0LL;
    
    long long l = 0LL;

    long long execints = 0LL;

    int n;                                                                   //1
    execints++;                                                                    
      
    std::cin >> n;                                                            //1
    execints++;

    std::vector<int> numbers(n);                                              //n
    execints += n; 
                                                       
    int max_index1 = -1;
    execints++; 

    execints++; 
    for (int i = 0; i < n; ++i) {                                             //1+2n
        
        execints += 2;                                             
        std::cin >> numbers[i];                                               //n
        
        execints++;                                               
    }

    execints++;  
    for(int i = 0; i < n; ++i){                                               //1 + 2n
        execints+=2;                       

        execints++;
        if((max_index1 == -1) || (numbers[i]>numbers[max_index1])){           //(n-1)n
            max_index1 = i;
            execints++;
            m++;                                                       
        }

        if ( (i <= 10) || (n - i <= 10)) std::cout << m << "\n";
    }

    int max_index2 = -1;                                                      //1
    for(int j = 0; j < n; ++j){                                               //1+2n
        execints+=2;                       

        execints++;                       
        if((j != max_index1) && ((max_index2 == -1) || (numbers[j]>numbers[max_index2]))){   //(2n-1)n
            max_index2 = j;
            execints++;
            l++;  
        }

        if ( (j <= 10) || (n - j <= 10)) std::cout << l << "\n";
    }

    std::cout <<((long long) (numbers[max_index1])*numbers[max_index2])<< "\n";         //3 + 2n + 2[(1+2n)(n-1)n] + (1+2n)
    return 0;

}

Overwriting max_pairwise_product3_super_execints.cpp


Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product3_super_execints.cpp -o max_pairwise_product3_super_execints
!./max_pairwise_product3_super_execints < input1.txt
!./max_pairwise_product3_super_execints < input2.txt
!./max_pairwise_product3_super_execints < input3.txt

1
2
3
4
5
6
7
8
9
10
1
2
3
4
5
6
7
8
9
9
90
1
2
3
3
3
3
1
2
2
3
4
4
40000000000
1
2
3
4
5
6
7
8
9
10
11
199991
199992
199993
199994
199995
199996
199997
199998
199999
200000
1
2
3
4
5
6
7
8
9
10
11
199991
199992
199993
199994
199995
199996
199997
199998
199999
199999
39999800000


## Python

Efficient cython file reading, string parsing, and array building

https://stackoverflow.com/questions/37219875/efficient-cython-file-reading-string-parsing-and-array-building



File PreProcessing With C and Cython

https://github.com/EveryTimeIWill18/Cython_Repo/blob/master/FastFileProcessingWithCython.ipynb



In [18]:
# python3

%%writefile main.py
import time

def mpp(numbers):
    n = len(numbers)
    maximum = 0                                              # 1
    for i in range(n):                                       # 1+2n
        for j in range(i + 1, n):                            # n+(n-1)n
            maximum = max(maximum, numbers[i] * numbers[j])  # 3
    return maximum

if __name__ == '__main__':
    input_n = int(input())
    list_numbers = [int(x) for x in input().split()]
    assert (len(list_numbers) == input_n)
    print(mpp(list_numbers))

Overwriting main.py


In [15]:
# python3

%%writefile main2.py
import time

def mpp_fast(n, numbers):
    max_i1 = 0                                      # 1
    max_i2 = 0                                      # 1
    for i in range(n):                              # 1+2n
        if (numbers[i] > numbers[max_i1]):          # n
            max_i2 = max_i1                         # 
            max_i1 = i                              # 2[1, n]
        if (numbers[i] > numbers[max_i2]):          # n
            max_i2 = i                              # [1, n]

    return numbers[max_i1] * numbers[max_i2]

if __name__ == '__main__':
    input_n = int(input())
    list_numbers = [int(x) for x in input().split()]
    assert (len(list_numbers) == input_n)
    print(mpp_fast(input_n, list_numbers))

Overwriting main2.py


The mpp_fast(n, numbers) method takes $$1+1+(1+2n)+n+(2[1,n])+n+([1,n]) = 3+4n+3[1,n]$$  steps to complete all the processing that means $3+7n$ in the worst case meanwhile the mpp(numbers) method takes $$1+(1+2n)+(n+(n-1)n)+3 = 5+2n+n^2$$ which is  $2-5n+n^2$ times slower than the fast algorithm

In [26]:
import random

n = 20000
l = []
max1 = -1
max2 = -1

for i in range(n):
  x = random.randint(0, n)
  l.append(str(x))
  if x >= max1:
    max2 = max1
    max1 = x
  elif x >= max2:
    max2 = x

fin = open("test2.in", "w")
fout = open("test2.out", "w")

fin.write(str(n) + "\n" + " ".join(l))
fout.write(str(max1 * max2))

fin.close()
fout.close()

In [27]:
%%time

!python main.py < test2.in > result.out
!grep -xvFf test2.out result.out

CPU times: user 368 ms, sys: 52.3 ms, total: 420 ms
Wall time: 52 s


In [28]:
%%time

!python main2.py < test2.in > result.out
!grep -xvFf test2.out result.out

400000000
CPU times: user 10.5 ms, sys: 13.5 ms, total: 24 ms
Wall time: 233 ms


# Screenhsoots of the Coursera Submissions in C++ and Python

![AplasB](https://raw.githubusercontent.com/gjhernandezp/algorithms/master/Lab6/AplusB.jpg)